src/blockchain/ain/jellyfish/utils/raw-tx-check.ts

```diff
@@ -0,0 +1,180 @@
+import { CTransactionSegWit, OP_DEFI_TX, Vin, Vout } from '@defichain/jellyfish-transaction';
+import { Config } from 'src/config/config';
+import { RawTxDto } from '../dto/raw-tx.dto';
+import { RawTxUtil } from './raw-tx-util';
+import { SmartBuffer } from 'smart-buffer';
+import { MasternodeOwnerService } from 'src/integration/masternode/application/services/masternode-owner.service';
+import { VaultService } from 'src/integration/vault/application/services/vault.service';
+import { TransactionType } from 'src/integration/transaction/domain/enums';
+
+export class RawTxCheck {
+  static incomingTxTypes = [
+    'OP_DEFI_TX_CREATE_MASTER_NODE',
+    'OP_DEFI_TX_RESIGN_MASTER_NODE',
+    'OP_DEFI_TX_VOTE',
+    'OP_DEFI_TX_CREATE_VAULT',
+    'OP_DEFI_TX_DEPOSIT_TO_VAULT',
+    'OP_DEFI_TX_TAKE_LOAN',
+    'OP_DEFI_TX_POOL_ADD_LIQUIDITY',
+    'OP_DEFI_TX_PAYBACK_LOAN',
+    'OP_DEFI_TX_WITHDRAW_FROM_VAULT',
+    'OP_DEFI_TX_POOL_REMOVE_LIQUIDITY',
+    'OP_DEFI_TX_COMPOSITE_SWAP',
+    'OP_DEFI_TX_ANY_ACCOUNT_TO_ACCOUNT',
+  ];
+  static outgoingTxTypes = ['OP_DEFI_TX_ANY_ACCOUNT_TO_ACCOUNT'];
+
+  static isAllowed(rawTx: RawTxDto, payload: any, listOfVaults: string[] = []): boolean {
+    const [, vouts] = this.parseInAndOutputs(rawTx.hex);
+    if (vouts.length === 0) return false;
+    const expectedIsIncoming = this.getExpectedIsIncoming(payload.type);
+    const [isIncoming, defiTxType] = this.isIncomingBasedOn(vouts, listOfVaults);
+    return (
+      (expectedIsIncoming !== undefined ? expectedIsIncoming === isIncoming : true) &&
+      this.isOnCorrectList(defiTxType, isIncoming ? this.incomingTxTypes : this.outgoingTxTypes)
+    );
+  }
```

Comment:
I wouldn't check the payload here. It's a RawTxCheck, nothing about payload 😉 so maybe like this (with using an enum of course):
```ts
static isAllowed(rawTx: RawTxDto, listOfVaults: string[] = []): 'Incoming' | 'Outgoing' | undefined {
    const [, vouts] = this.parseInAndOutputs(rawTx.hex);
    if (vouts.length === 0) return false;

    const [direction, defiTxType] = this.getDirection(vouts, listOfVaults); // returns Incoming | Outgoing
    return this.isOnCorrectList(defiTxType, direction === 'Incoming' ? this.incomingTxTypes : this.outgoingTxTypes) ? direction : undefined;
  }
```
So the method would just return the direction, in which the transaction is going (and undefined, if it's not allowed, or even through directly).
Then we could actually set this to the payload in transaction-execution.service, instead of doing this twice 🤔

Reply:
Or let's not place it in the payload, as we anyway do not use it in CW. Maybe just add a new parameter to the transaction.service sign method?